spring-integration-core/src/main/java/org/springframework/integration/aggregator/AbstractCorrelatingMessageHandler.java

```diff
@@ -753,10 +755,24 @@ protected void completeGroup(Object correlationKey, MessageGroup group) {
 		Object result = this.outputProcessor.processMessageGroup(group);
 		Collection<Message<?>> partialSequence = null;
 		if (result instanceof Collection<?>) {
-			this.verifyResultCollectionConsistsOfMessages((Collection<?>) result);
+			verifyResultCollectionConsistsOfMessages((Collection<?>) result);
 			partialSequence = (Collection<Message<?>>) result;
 		}
-		this.sendOutputs(result, message);
+
+		if (partialSequence == null && !(result instanceof Message<?>)) {
+			AbstractIntegrationMessageBuilder<?> messageBuilder;
+			if (result instanceof AbstractIntegrationMessageBuilder<?>) {
+				messageBuilder = (AbstractIntegrationMessageBuilder<?>) result;
+			}
+			else {
+				messageBuilder = getMessageBuilderFactory()
+						.withPayload(result)
+						.copyHeaders(message.getHeaders());
+			}
+			result = messageBuilder.popSequenceDetails();
+		}
```

Comment:
Now I look at this, I wonder if it is completely valid to do this unconditionally. Consider

    splitter -> custom-splitter -> custom-aggregator -> aggregator

Perhaps we need a new flag `alwaysPopSequenceDetails` (true in 5.1, false in 5.0).

Or, add some metadata (`seqInfoId`)  to the sequence details and only pop if the splitter and aggregator are configured with the same metadata.

If the custom splitter doesn't push sequence details, we will remove the real ones from the first splitter.

Reply:
OK. Sounds like we just need to be sure that our aggregation is based on the `sequenceDetails`. Because custom `splitter` may not be based on the `applySequence`.
So, maybe just an extra `sequenceApplied` header? But then how to understand on which level it has not been used?..

Or... I just don't see the whole picture and need to sleep with this...

Reply:
After sleeping with this, I came with the conclusion that your suggestion about `popSequenceDetails(true|false)` is the way to go.
We won't pollute messages with some extra header which may be really out of control and I don't want to tie an aggregator with the splitter.

We may do a condition based on the default `HeaderAttributeCorrelationStrategy(IntegrationMessageHeaderAccessor.CORRELATION_ID)` and `SimpleSequenceSizeReleaseStrategy`, but that still doesn't come as a robust solution since we can use defaults, but populate sequence header some other way, not via splitter.

From here I think we just need to add such a `popSequenceDetails` option to the correlation handler, make it `true` by default (looks like most cases, since we haven't had issues so far), modify a `AbstractAggregatingMessageGroupProcessor` to return a `AbstractIntegrationMessageBuilder` instead of the whole message and do conditional `.popSequenceDetails()` already in the `AbstractCorrelatingMessageHandler`. 

Saying that I don't think that we are good for back-porting to `5.0.x`: looks like this is going to be some breaking change.

Let me know WDYT?
Thanks

Reply:
Maybe call it simply `popSequence` with the documentation explaining it undoes the nearest upstream `applySequence`?

Reply:
Makes sense. Will fix soon.